jwst/saturation/saturation.py

```diff
@@ -13,89 +13,103 @@
-    #   and saturation mask
-    wh_sat = np.bitwise_and(dqmask, dqflags.pixel['NO_SAT_CHECK'])
-    dqmask[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = dqflags.pixel['NO_SAT_CHECK']
-    satmask[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = HUGE_NUM
-    # Correct saturation values for NaNs in the ref file
-    correct_for_NaN(satmask, dqmask)
-
-    dq_flag = dqflags.group['SATURATED']
+    # For pixels flagged in reference file as NO_SAT_CHECK,
+    # set the output dq mask to NO_SAT_CHECK and set the
+    # saturation check threshold to the A-to-D converter limit.
+    wh_sat = np.bitwise_and(sat_dq, dqflags.pixel['NO_SAT_CHECK'])
+    sat_dq[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = dqflags.pixel['NO_SAT_CHECK']
+    sat_thresh[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = ATOD_LIMIT
+    del wh_sat
 
-    nints = ramparr.shape[0]
-    ngroups = ramparr.shape[1]
+    # Reset NaN values in the saturation threshold array
+    correct_for_NaN(sat_thresh, sat_dq)
 
-    detector = input_model.meta.instrument.detector
-    flagarray = np.zeros(ramparr.shape[-2:], dtype=groupdq.dtype)
+    # Loop over integrations and groups, checking for pixel values
+    # that are above the saturation threshold
+    flagarray = np.zeros(ramp_array.shape[-2:], dtype=groupdq.dtype)
     for ints in range(nints):
-        for plane in range(ngroups):
+        for group in range(ngroups):
             # Update the 4D groupdq array with the saturation flag. The
-            # flag is set in the current plane and all following planes.
+            # flag is set in the current group and all following groups.
             if is_irs2_format:
-                sci_temp = x_irs2.from_irs2(ramparr[ints, plane, :, :],
-                                            irs2_mask, detector)
-                flag_temp = np.where(sci_temp >= satmask, dq_flag, 0)
+                sci_temp = x_irs2.from_irs2(ramp_array[ints, group, :, :],
+                                            irs2_mask,
+                                            detector)
+                flag_temp = np.where(sci_temp >= sat_thresh, sat_flag, 0)
                 # Copy flag_temp into flagarray.
                 x_irs2.to_irs2(flagarray, flag_temp, irs2_mask, detector)
+
             else:
-                flagarray[:, :] = np.where(ramparr[ints, plane, :, :] >= satmask,
-                                          dq_flag, 0)
-            np.bitwise_or(groupdq[ints, plane:, :, :], flagarray,
-                          groupdq[ints, plane:, :, :])
+                flagarray[:, :] = np.where(ramp_array[ints, group, :, :] >= sat_thresh,
+                                           sat_flag, 0)
+            np.bitwise_or(groupdq[ints, group:, :, :], flagarray,
+                          groupdq[ints, group:, :, :])
 
+    # Save the saturation flags in the output GROUPDQ array
     output_model.groupdq = groupdq
+
+    n_sat = len(np.where(np.any(np.any(groupdq, axis=0), axis=0))[0])
```

Comment:
Any of you numpy ninjas out there have a better, more concise way to do this counting of saturated pixels? I just want the number of pixels flagged as saturated in at least 1 integration and 1 group (i.e. I do NOT want to have multiple counts due to the pixel being saturated in more than 1 group or more than 1 integration).

Reply:
Whoops. Actually that's the wrong computation. It's counting any pixel that has a non-zero flag value in the groupdq array. I only want ones that have the SATURATION flag. Stay tuned for next commit ...